libraries/helpers.rb

```diff
@@ -73,5 +73,27 @@ def home_basedir
     def pubkey_type(pubkey)
       %w(ed25519 ecdsa dss rsa dsa).filter { |kt| pubkey.split.first.include?(kt) }.first || 'rsa'
     end
+
+    def primary_gid(user)
```

Comment:
This is very unclear, too many conditions and use of `||`. Off the cuff it could be refactored into something like below but preferably without the ORs as they make it very messy and confusing.

```ruby
def primary_gid(user)
  return user[:id] || user[:username] unless user.key?(:gid) && !user[:gid].nil?

  user[:primary_group] || user[:gid].to_i
end
```